internal/pkg/util/fs/overlay/overlay_item_linux.go

```diff
@@ -0,0 +1,297 @@
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package overlay
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/util/bin"
+	"github.com/sylabs/singularity/pkg/image"
+	"github.com/sylabs/singularity/pkg/sylog"
+)
+
+// Item represents information about a single overlay item (as specified,
+// for example, in a single --overlay argument)
+type Item struct {
+	// Type represents what type of overlay this is (from among the values in pkg/image)
+	Type int
+
+	// Writable represents whether this is a writable overlay
+	Writable bool
+
+	// BarePath is the path of the overlay stripped of any colon-prefixed
+	// options (like ":ro")
+	BarePath string
+
+	// DirToMount is the path of the directory that will actually be passed to
+	// the mount system-call when mounting this overlay
+	DirToMount string
+
+	// secureParentDir is the (optional) location of a secure parent-directory
+	// in which to create mount directories if needed. If empty, one will be
+	// created with os.MkdirTemp()
+	secureParentDir string
+}
+
+// NewOverlayFromString takes a string argument, as passed to --overlay, and returns
+// an overlayInfo struct describing the requested overlay.
+func NewOverlayFromString(overlayString string) (*Item, error) {
+	overlay := Item{Writable: true}
+
+	splitted := strings.SplitN(overlayString, ":", 2)
+	overlay.BarePath = splitted[0]
+	if len(splitted) > 1 {
+		if splitted[1] == "ro" {
+			overlay.Writable = false
+		}
+	}
+
+	s, err := os.Stat(overlay.BarePath)
+	if (err != nil) && os.IsNotExist(err) {
+		return nil, fmt.Errorf("specified overlay %q does not exist", overlay.BarePath)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if s.IsDir() {
+		overlay.Type = image.SANDBOX
+	} else if err := analyzeImageFile(&overlay); err != nil {
+		return nil, fmt.Errorf("error encountered while examining image file %s: %s", overlay.BarePath, err)
+	}
+
+	return &overlay, nil
+}
+
+// analyzeImageFile attempts to determine the format of an image file based on
+// its header
+func analyzeImageFile(overlay *Item) error {
+	img, err := image.Init(overlay.BarePath, overlay.Writable)
+	if err != nil {
+		return fmt.Errorf("error encountered while trying to examine image: %s", err)
+	}
+
+	switch img.Type {
+	case image.SQUASHFS:
+		overlay.Type = image.SQUASHFS
+		// squashfs image must be readonly
+		overlay.Writable = false
+		return nil
+	case image.EXT3:
+		overlay.Type = image.EXT3
+	default:
+		return fmt.Errorf("image %s is of a type that is not currently supported as overlay", overlay.BarePath)
+	}
+
+	return nil
+}
+
+// SetSecureParentDir sets the secure parent-dir in which to create
+// overlay-specific mount directories.
+func (i *Item) SetSecureParentDir(d string) {
+	i.secureParentDir = d
```

Comment:
If we're calling it `secureParentDir`, and letting it be set from a public method, I'd anticipate some checks on it when set to ensure it is secure.

Otherwise, if we don't want to enforce secure permissions here, it's just a `parentDir`.